### PR TITLE
Database RW Split

### PR DESCRIPTION
**Pull request by LBeernaertProton**
Ensure all the database queries are accessed through the scope of RWLock
    to avoid the infamous "database locked" error and to allow more parallel
    read only requests from clients.
    
 The RWLock guarantees that only one goroutine/thread can write to the
 datbase, therefore ensuring we never trigger the limitations of SQLite
 DB which causes the error.
    
 The removal of the global TX Lock now enables clients to perform more
 read only request in parallel.
    
 Going forward it is expected all database access to be accessed via the
 `DB` type.

### CODE REVIEW

File: internal/backend/db_mailbox.go [hunk 1]
```diff
@@ -166,7 +166,7 @@ func DBGetMailboxMessageIDPairs(ctx context.Context, client *ent.Client, mailbox
 	}), nil
 }
 
-func DBGetUIDInterval(ctx context.Context, mbox *ent.Mailbox, begin, end int) ([]*ent.UID, error) {
+func DBGetUIDInterval(ctx context.Context, client *ent.Client, mbox *ent.Mailbox, begin, end int) ([]*ent.UID, error) {
```
Comment (jameshoulahan) [inline internal/backend/db_mailbox.go:169]: need client here?

File: internal/backend/db_mailbox.go [hunk 2]
```diff
@@ -211,15 +211,15 @@ func DBGetMailboxByID(ctx context.Context, client *ent.Client, id imap.InternalM
 	return client.Mailbox.Query().Where(mailbox.MailboxID(id)).Only(ctx)
 }
 
-func DBGetMailboxMessages(ctx context.Context, mbox *ent.Mailbox) ([]*ent.UID, error) {
+func DBGetMailboxMessages(ctx context.Context, client *ent.Client, mbox *ent.Mailbox) ([]*ent.UID, error) {
```
Comment (jameshoulahan) [inline internal/backend/db_mailbox.go:214]: need client here too?

File: internal/backend/db_mailbox.go (hunk 1 above)
Comment (LBeernaertProton) [reply internal/backend/db_mailbox.go:169]: While we don't need to use the client object, it is there to enforce the DB type use and to make sure we have acquire the lock. Otherwise, these queries will run at any moment, bypassing the lock and causing the infamous "database locked" error.

File: internal/backend/db_mailbox.go [hunk 3]
```diff
@@ -211,15 +211,15 @@ func DBGetMailboxByID(ctx context.Context, client *ent.Client, id imap.InternalM
...
 	return mbox.QueryUIDs().WithMessage().All(ctx)
 }
 
 func DBGetMailboxRecentCount(ctx context.Context, mbox *ent.Mailbox) (int, error) {
 	return mbox.QueryUIDs().Where(uid.Recent(true)).Count(ctx)
 }
 
-func DBGetMailboxMessagesForNewSnapshot(ctx context.Context, mbox *ent.Mailbox) ([]*ent.UID, error) {
+func DBGetMailboxMessagesForNewSnapshot(ctx context.Context, client *ent.Client, mbox *ent.Mailbox) ([]*ent.UID, error) {
```
Comment (jameshoulahan) [inline internal/backend/db_mailbox.go:222]: or here?

Comment (LBeernaertProton) [reply internal/backend/db_mailbox.go:222]: While we don't need to use the client object, it is there to enforce the DB type use and to make sure we have acquire the lock. Otherwise, these queries will run at any moment, bypassing the lock and causing the infamous "database locked" error.

File: internal/backend/db.go [hunk 4]
```diff
@@ -24,19 +24,41 @@ func (d *DB) Init(ctx context.Context) error {
 }
 
 func (d *DB) Read(ctx context.Context, fn func(context.Context, *ent.Client) error) error {
-	d.lock.RLock()
-	defer d.lock.Unlock()
+	_, err := DBReadResult[struct{}](ctx, d, func(ctx context.Context, client *ent.Client) (struct{}, error) {
```
Comment (jameshoulahan) [inline internal/backend/db.go:27]:
```suggestion
	_, err := DBReadResult(ctx, d, func(ctx context.Context, client *ent.Client) (struct{}, error) {
```

File: internal/backend/db.go [hunk 5]
```diff
@@ -24,19 +24,41 @@ func (d *DB) Init(ctx context.Context) error {
...
+	})
 
-	return fn(ctx, d.db)
+	return err
 }
 
 func (d *DB) Write(ctx context.Context, fn func(context.Context, *ent.Tx) error) error {
-	d.lock.Lock()
-	defer d.lock.Unlock()
+	_, err := DBWriteResult[struct{}](ctx, d, func(ctx context.Context, tx *ent.Tx) (struct{}, error) {
```
Comment (jameshoulahan) [inline internal/backend/db.go:35]:
```suggestion
	_, err := DBWriteResult(ctx, d, func(ctx context.Context, tx *ent.Tx) (struct{}, error) {
```

Comment (LBeernaertProton) [reply internal/backend/db.go]: Done

File: internal/backend/db.go (hunk 4 above)
Comment (LBeernaertProton) [reply internal/backend/db.go]: Done

File: internal/backend/user.go [hunk 6]
```diff
@@ -53,12 +53,12 @@ func newUser(ctx context.Context, userID string, db *DB, remote *remote.User, st
...
+				update := func() error {
 					defer update.Done()
-					return user.apply(context.Background(), tx, update)
-				}); err != nil {
-					logrus.WithError(err).Errorf("Failed to apply update: %v", update)
+					return user.apply(context.Background(), update)
+				}
+				if err := update(); err != nil {
+					logrus.WithError(err).Errorf("Failed to apply update: %v", err)
 				}
```
Comment (jameshoulahan) [inline internal/backend/user.go:62]: now that `user.apply` is only used for external updates, we should probably just move this deferred `update.Done()` call into `user.apply` itself. That would simplify this weirdness here.

+ the docstring on `user.apply` is out of date, there's no `source` anymore.

File: internal/backend/user_updates.go [hunk 7]
```diff
@@ -15,89 +15,100 @@ import (
...
 // applyMailboxDeleted applies a MailboxDeleted update.
-func (user *user) applyMailboxDeleted(ctx context.Context, tx *ent.Tx, update *imap.MailboxDeleted) error {
-	if exists, err := DBMailboxExistsWithRemoteID(ctx, tx.Client(), update.MailboxID); err != nil {
+func (user *user) applyMailboxDeleted(ctx context.Context, update *imap.MailboxDeleted) error {
+	if exists, err := DBReadResult(ctx, user.db, func(ctx context.Context, client *ent.Client) (bool, error) {
+		return DBMailboxExistsWithRemoteID(ctx, client, update.MailboxID)
+	}); err != nil {
 		return err
-	} else if !exists {
+	} else if exists {
```
Comment (jameshoulahan) [inline internal/backend/user_updates.go:90]: still in need of a decent morning coffee, but why did you change this from `!exists` to `exists`...? was this a bug you fixed, or did you make a mistake here?

File: internal/backend/user_updates.go [hunk 8]
```diff
@@ -15,89 +15,100 @@ import (
...
-func (user *user) applyMailboxUpdated(ctx context.Context, tx *ent.Tx, update *imap.MailboxUpdated) error {
-	client := tx.Client()
-	if exists, err := DBMailboxExistsWithRemoteID(ctx, client, update.MailboxID); err != nil {
+func (user *user) applyMailboxUpdated(ctx context.Context, update *imap.MailboxUpdated) error {
+	if exists, err := DBReadResult(ctx, user.db, func(ctx context.Context, client *ent.Client) (bool, error) {
+		return DBMailboxExistsWithRemoteID(ctx, client, update.MailboxID)
+	}); err != nil {
 		return err
-	} else if !exists {
+	} else if exists {
```
Comment (jameshoulahan) [inline internal/backend/user_updates.go:105]: same here re `!exists` -> `exists`

File: internal/session/handle.go [hunk 9]
```diff
@@ -12,7 +12,7 @@ import (
 )
 
 func (s *Session) handleOther(ctx context.Context, tag string, cmd *proto.Command, profiler profiling.CmdProfiler) chan response.Response {
-	ch := make(chan response.Response)
+	ch := make(chan response.Response, 100)
```
Comment (jameshoulahan) [inline internal/session/handle.go:15]: why buffer this channel? and how'd you arrive at 100?

Comment (LBeernaertProton) [reply internal/session/handle.go:15]: If you don't buffer the channel, you can only send one item and then you need to wait for that one item to be read. If it takes some time to send the reply from gluon to the client, the command code can't proceed until gluon fetches the next response. Buffering now allows the command to execute and release all the locks sooner as it can queue up N response items without waiting for them to be sent.

The number 100 is arbitrary, I was considering large fetch request for this case.

File: internal/backend/user_updates.go (hunk 7 above)
Comment (LBeernaertProton) [reply internal/backend/user_updates.go]: Good catch. Fixed.

File: internal/backend/user_updates.go (hunk 8 above)
Comment (LBeernaertProton) [reply internal/backend/user_updates.go]: Good catch. Fixed.

File: internal/backend/user.go (hunk 6 above)
Comment (LBeernaertProton) [reply internal/backend/user.go]: Fixed.

File: internal/session/handle.go (hunk 9 above)
Comment (jameshoulahan) [reply internal/session/handle.go:15]: Good point, nice change.

In that case, i'd like the 100 to be defined as a package constant, potentially even a variable such that it can be configured (in future, not in this MR) by a gluon init option.

**Review by jameshoulahan — APPROVED** [review]
looks good, just make sure to define the magic number `100` as a const or var in `internal/session/init.go` in a followup PR
